src/core/prototypes/state.ts

```diff
@@ -72,6 +85,209 @@ export class ChartStateManager {
     }
   }
 
+  public resetDifference() {
+    this.chartOrigin = deepClone(this.chart);
+  }
+
+  // eslint-disable-next-line max-lines-per-function
+  public hasUnsavedChanges() {
+    const origin = this.chartOrigin;
+    const chart = this.chart;
+
+    const currentProperties = chart.properties;
+    const originProperties = origin.properties;
+
+    try {
+      expect_deep_approximately_equals(currentProperties, originProperties, 0);
+    } catch {
+      return true;
+    }
+
+    if (origin.constraints.length !== chart.constraints.length) {
+      return true;
+    } else {
+      for (let index = 0; index < origin.constraints.length; index++) {
```

Comment:
Can you comment on why this section below isn't caught by `expect_deep_approximately_equals`?